core/src/unprocessed_packet_batches.rs

```diff
@@ -79,6 +90,50 @@ impl UnprocessedPacketBatches {
     pub fn with_capacity(capacity: usize) -> Self {
         UnprocessedPacketBatches(VecDeque::with_capacity(capacity))
     }
+
+    /// Returns total number of all packets (including unprocessed and processed) in buffer
+    #[allow(dead_code)]
+    fn get_packets_count(&self) -> usize {
+        self.iter()
+            .map(|deserialized_packet_batch| deserialized_packet_batch.packet_batch.packets.len())
+            .sum()
+    }
+
+    /// Returns total number of unprocessed packets in buffer
+    #[allow(dead_code)]
+    fn get_unprocessed_packets_count(&self) -> usize {
+        self.iter()
+            .map(|deserialized_packet_batch| deserialized_packet_batch.unprocessed_packets.len())
+            .sum()
+    }
+
+    /// Iterates packets in buffered packet_batches, returns all unprocessed packet's stake,
+    /// and its locator
+    #[allow(dead_code)]
+    fn get_stakes_and_locators(&self) -> (Vec<u64>, Vec<PacketLocator>) {
+        let num_unprocessed_packets = self.get_unprocessed_packets_count();
+        let mut stakes = Vec::<u64>::with_capacity(num_unprocessed_packets);
+        let mut locators = Vec::<PacketLocator>::with_capacity(num_unprocessed_packets);
+
+        self.iter()
+            .enumerate()
+            .for_each(|(batch_index, deserialized_packet_batch)| {
+                let packet_batch = &deserialized_packet_batch.packet_batch;
+                deserialized_packet_batch
+                    .unprocessed_packets
+                    .keys()
+                    .for_each(|packet_index| {
+                        let p = &packet_batch.packets[*packet_index];
+                        stakes.push(p.meta.sender_stake);
+                        locators.push(PacketLocator {
+                            batch_index,
+                            packet_index: *packet_index,
+                        });
+                    });
+            });
+
+        (stakes, locators)
+    }
```

Comment:
I think this can be simplified to:

```
self.iter()
            .enumerate()
            .flat_map(|(batch_index, deserialized_packet_batch)| {
                let packet_batch = &deserialized_packet_batch.packet_batch;
                deserialized_packet_batch
                    .unprocessed_packets
                    .keys()
                    .map(move |packet_index| {
                        let p = &packet_batch.packets[*packet_index];
                        (
                            p.meta.sender_stake,
                            PacketLocator {
                                batch_index,
                                packet_index: *packet_index,
                            },
                        )
                    })
            })
            .unzip()
```

Reply:
Was hoping `with_capacity` saves a bit from memory allocation. But `flat_map` does make the code much readable. Thanks for suggestions, also addresses Trent's comment.